pkg/controller/download_request_controller.go

```diff
@@ -18,229 +18,136 @@ package controller
-	return errors.WithStack(err)
-}
-
-// deleteIfExpired deletes downloadRequest if it has expired.
-func (c *downloadRequestController) deleteIfExpired(downloadRequest *velerov1api.DownloadRequest) error {
-	log := c.logger.WithField("key", kube.NamespaceAndName(downloadRequest))
-	log.Info("checking for expiration of DownloadRequest")
-	if downloadRequest.Status.Expiration.Time.After(c.clock.Now()) {
-		log.Debug("DownloadRequest has not expired")
-		return nil
-	}
+		backup := &velerov1api.Backup{}
+		if err := r.Client.Get(r.Ctx, kbclient.ObjectKey{
+			Namespace: downloadRequest.Namespace,
+			Name:      backupName,
+		}, backup); err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
 
-	log.Debug("DownloadRequest has expired - deleting")
-	return errors.WithStack(c.downloadRequestClient.DownloadRequests(downloadRequest.Namespace).Delete(context.TODO(), downloadRequest.Name, metav1.DeleteOptions{}))
-}
+		location := &velerov1api.BackupStorageLocation{}
+		if err := r.Client.Get(r.Ctx, kbclient.ObjectKey{
+			Namespace: backup.Namespace,
+			Name:      backup.Spec.StorageLocation,
+		}, location); err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
 
-// resync requeues all the DownloadRequests in the lister's cache. This is mostly to handle deleting
-// any expired requests that were not deleted as part of the normal client flow for whatever reason.
-func (c *downloadRequestController) resync() {
-	list, err := c.downloadRequestLister.List(labels.Everything())
-	if err != nil {
-		c.logger.WithError(errors.WithStack(err)).Error("error listing download requests")
-		return
-	}
+		pluginManager := r.NewPluginManager(log)
+		defer pluginManager.CleanupClients()
 
-	for _, dr := range list {
-		key, err := cache.MetaNamespaceKeyFunc(dr)
+		backupStore, err := r.NewBackupStore(location, pluginManager, log)
 		if err != nil {
-			c.logger.WithError(errors.WithStack(err)).WithField(DownloadRequest, dr.Name).Error("error generating key for download request")
-			continue
+			log.WithError(err).Error("Error getting a backup store")
+			return ctrl.Result{}, errors.WithStack(err)
 		}
 
-		c.queue.Add(key)
-	}
-}
-
-func patchDownloadRequest(original, updated *velerov1api.DownloadRequest, client velerov1client.DownloadRequestsGetter) (*velerov1api.DownloadRequest, error) {
-	origBytes, err := json.Marshal(original)
-	if err != nil {
-		return nil, errors.Wrap(err, "error marshalling original download request")
-	}
-
-	updatedBytes, err := json.Marshal(updated)
-	if err != nil {
-		return nil, errors.Wrap(err, "error marshalling updated download request")
-	}
-
-	patchBytes, err := jsonpatch.CreateMergePatch(origBytes, updatedBytes)
-	if err != nil {
-		return nil, errors.Wrap(err, "error creating json merge patch for download request")
+		if downloadRequest.Status.DownloadURL, err = backupStore.GetDownloadURL(downloadRequest.Spec.Target); err != nil {
```

Comment:
If we never find a DownloadURL here, will it just constantly re-queue?

Maybe we should check for expiration at the beginning, regardless of whether the downloadrequest.Status.Phase field is processed, so that erroneous requests don't sit and spin in the controller forever, filling up the logs with errors. Since DownloadRequests are usually a synchronous activity where someone tries to download a thing at the command line, I'm not sure if it's worth continuing to try past the TTL. Does that sound reasonable to others?

Reply:
Right now it looks like the expiration is only set once a DownloadURL has been received (which happens in the case of success, we have a DownloadURL which expires after TTL). Are you suggesting that we should instead set the expiration at the point where we **create** the DownloadRequest, and then the expiration of the request would be tied to time since creation, rather than the expiration time of the resulting URL?

Reply:
Addressed this as per Nolan's suggestion.